Memoize the result of config.get_config

alalazo opened the pull request:
fixes #19539

Profiling a few common commands (`spack spec <spec>`, `spack install <spec>`) it seems we are spending a lot of time merging YAML files again and again.

Almost all of the calls are from `get_config`, thus to speed things up we can "just" employ a cache.

alalazo committed:
Memoize the result of config.get_config

fixes #19539

Profiling a few common commands (spack spec <spec>,
spack install <spec>) it seems we are spending a lot
of time merging YAML files again and again.

Almost all of the calls are from `get_config`, thus
to speed things up we can "just" employ a cache.

alalazo commented:
@haampie

scheibelp commented:
That `deepcopy` was added so that changing the merged yaml wouldn't change the underlying scope that the config was merged from.

In this case, I'm concerned that the memoization might have the opposite effect: if the user does something like

```
v1 = get_config(x, scope=None)
v2 = get_config(x, scope='user')
# user changes v2
v3 = get_config(x, scope=None)
```

I think `v3` should reflect the changes to `v2`. With memoization, I don't think it would.

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -498,6 +499,9 @@ def update_config(self, section, update_data, scope=None, force=False):
             force (str): force the update
         """
+        # Invalidate the cache
+        self._get_config_memoized.cache.clear()
```
@scheibelp I share your concern. This is a lot faster than before. @haampie reported his use case going:
- From 11min. before #18482 
- To 90 min. with #18482
- To 4 min. with this PR

Unit tests are running 10-15 mins. faster than other recent PRs and I see speedups of 4x on simple `spack spec ...` commands. I'm already trying to intercept writes to invalidate the cache. I agree though that if it works, it would feel like taking a very fast shortcut through a minefield...

alalazo replied:
> I think v3 should reflect the changes to v2

How does the user change v2 in the configuration? If the user goes through the API it should work if we clear the cache on commands that mutate the state of the config.

scheibelp replied:
> I'm already trying to intercept writes to invalidate the cache

I'm curious what your approach is.

It's possible things could be simplified by adding a `read_only` option to `get_config`. Most parts of the code retrieve the configuration to read it and in that case a `deepcopy` should not be required. I think I'd rather trust these parts of the code for the time being and add an issue to do the interception you suggest (which I think would address both the modification and speed concerns).

alalazo replied:
> It's possible things could be simplified by adding a read_only option to get_config

This is not without drawbacks as it would make the API signature harder to understand. Can you clarify how a user would mutate v2 in your example above? I'm trying to understand if it is expected to go through `set_config` or other methods of the `Configuration` class.

scheibelp replied:
>  Can you clarify how a user would mutate v2 in your example above?

Config objects behave like nested dictionaries, so I think things like

```
v2['foo'] = 'bar'
```

would be fair game. That being said I'm wondering if we generally wrap up all our modifications with `update_config` or `config.set` - invalidating the cache in those methods might be sufficient. That would require us to agree on updating the API and documentation for `Config` to reflect this requirement.

scheibelp replied:
> it would make the API signature harder to understand

Adding a cache that has to be cleared in disperse places (i.e. the locations you have added `_get_config_memoized_cache.clear()`) also complicates the API. I'd rather avoid that if possible. 

That could be simplified in your approach here by moving those calls into the `config` module. For example: never set `spack.config.config` but rather call a function like `spack.config.set_config`; `config.remove_scope` could also clear the cache automatically.

alalazo replied:
> That being said I'm wondering [ ... ] That would require us to agree on updating the API and documentation for Config to reflect this requirement.

That's what I was thinking too. If we agree to go through those methods we have a much smaller surface to check. I think it's safe to say that unit tests are surely going through `Configuration` methods.

alalazo replied:
> Adding a cache that has to be cleared in disperse places (i.e. the locations you have added _get_config_memoized_cache.clear()) also complicates the API. I'd rather avoid that if possible.

Ugh, I didn't get a refresh automatically, so I saw your second comment just now. I did something that goes in the same direction with decorators in 27e0f32 If that looks good enough I'll proceed changing docstrings and maybe docs.

alalazo committed:
Use a decorator to be more declarative

This also factors into a single place the
call to clear the memoization cache.

scheibelp commented on lib/spack/spack/test/conftest.py:
```diff
@@ -318,6 +318,7 @@ def _skip_if_missing_executables(request):
 @contextlib.contextmanager
 def use_configuration(config):
     """Context manager to swap out the global Spack configuration."""
+    config._get_config_memoized.cache.clear()
```
I think we should also discourage use of `spack.config.config = ...` and add a method to explicitly set it in `spack.config`. that method should be called twice here:

```
spack.config.config = config
...
spack.config.config = saved
```

(either way, I think you'll want to clear the memoizations twice)

alalazo replied:
Added `spack.config.replace_config` in 6dcb935

scheibelp commented on lib/spack/spack/test/cmd/common/arguments.py:
```diff
@@ -20,6 +20,7 @@ def parser():
     arguments.add_common_arguments(p, ['jobs'])
     yield p
     # Cleanup the command line scope if it was set during tests
+    spack.config.config._get_config_memoized.cache.clear()
```
I think any test which sets or changes a scope should be forced to use a `spack.config` module function (which should clear memoizations). I think for example this could just call `spack.config.config.clear_caches()`

alalazo replied:
Done.

alalazo committed:
"replace_config": explicitly replace the global configuration

Added a function in spack.config to swap
configurations explicitly. That function
takes care of implementation details,
like clearing caches.

alalazo commented:
@scheibelp Let me know if you think anything else is needed here.

scheibelp commented:
I'm looking through this again and I think this works, but I want to make it clearer to users that writes should be done with `Configuration.update_config`. I think it would be good to change the name of `ConfigScope.write_section` to `ConfigScope._write_section` (and to update the docstring to indicate that users of the `config` module should use `update_config`).

alalazo committed:
Prefix the write function of all the config scopes with an underscore

This is to discourage users from calling them directly

alalazo commented:
@scheibelp Done!

scheibelp reviewed (approved):
Suggested commit message: 

```
Config: cache results of get_config

`config.get_config` now caches the results and returns the same
configuration if called multiple times with the same arguments
(i.e. the same section and scope).

As a consequence, it is expected that users will always call
`update_config` after changing the configuration (even if
manipulating it as a Python nested dictionary). Failure to do
so will lead to unexpected results (e.g. calling get_config
after changing the configuration will not reflect the changes
since the first call to get_config).

Likewise, tests that need to change the global configuration should
use the newly-provided `config.replace_config` function.

Any modifications to the configuration should use one of these two
methods.
```

haampie commented:
@alalazo this breaks `InternalConfigScope` since it deletes values forever when "clearing cache"

alalazo commented:
> this breaks InternalConfigScope since it deletes values forever when "clearing cache"

@haampie I disagree. The `InternalConfigScope` implementation has been buggy since it was introduced, this PR just reveals the bug. It inherits from `ConfigScope` and that class has a `clear` method that should be honored. I fixed recently a similar bug in #22559

haampie commented:
hm yeah, in any case, a bug.

alalazo commented:
@haampie Can you make a separate report for it? Or I can do it if you prefer. Point is the bug has nothing to do with this PR imho.
